bugwarrior/services/logseq.py

```diff
@@ -0,0 +1,301 @@
+                scheduled_date = self.get_scheduled_date(line)
+            elif line.startswith("DEADLINE: "):
+                deadline_date = self.get_scheduled_date(line)
+            else:
+                annotations.append(self._unescape_content(line))
+        annotations.pop(0)  # remove first line
+        return annotations, scheduled_date, deadline_date
+
+    def get_url(self):
+        return f'logseq://graph/{self.extra["graph"]}?block-id={self.record["uuid"]}'
+
+    def get_logseq_state(self):
+        return self.record["marker"]
+
+    def get_scheduled_date(self, scheduled):
+        # format is <YYYY-MO-DD DAY HH:MM .+1d>
+        # e.g. <2024-06-20 Thu 10:55 .+1d>
+        date_split = (
+            scheduled.replace("DEADLINE: <", "")
+            .replace("SCHEDULED: <", "")
+            .replace(">", "")
+            .split(" ")
+        )
+        date = ""
+        date_format = ""
+        if len(date_split) == 2:  # <date day>
+            date = date_split[0]
+            date_format = "%Y-%m-%d"
+        elif len(date_split) == 3 and (date_split[2][0] in ("+", ".")):  # <date day repeat>
+            date = date_split[0]
+            date_format = "%Y-%m-%d"
+        elif len(date_split) == 3:  # <date day time>
+            date = date_split[0] + " " + date_split[2]
+            date_format = "%Y-%m-%d %H:%M"
+        elif len(date_split) == 4:  # <date date time repeat>
+            date = date_split[0] + " " + date_split[2]
+            date_format = "%Y-%m-%d %H:%M"
+        else:
+            log.warning(f"Could not determine date format from {scheduled}")
+            return None
+
+        try:
+            return datetime.strptime(date, date_format)
+        except ValueError:
+            log.warning(f"Could not parse date {date} from {scheduled}")
+        return None
+
+    def _is_waiting(self):
+        return self.STATE_MAP[self.get_logseq_state()] == "waiting"
+
+    def to_taskwarrior(self):
+        annotations, scheduled_date, deadline_date = self.get_annotations_from_content()
+        wait_date = min([d for d in [scheduled_date, deadline_date, self.SOMEDAY] if d is not None])
+        return {
+            "project": self.extra["graph"],
+            "priority": (
+                self.PRIORITY_MAP[self.record["priority"]]
+                if "priority" in self.record
+                else None
+            ),
+            "annotations": annotations,
+            "tags": self.get_tags_from_content(),
+            "due": deadline_date,
+            "scheduled": scheduled_date,
+            "wait": wait_date if self._is_waiting() else None,
+            "status": self.STATE_MAP[self.get_logseq_state()],
+            self.ID: self.record["id"],
+            self.UUID: self.record["uuid"],
+            self.STATE: self.record["marker"],
+            self.TITLE: self.get_formated_title(),
+            self.SCHEDULED: scheduled_date,
+            self.DEADLINE: deadline_date,
```

Comment:
Why do we need these UDA's when the same data is already in `scheduled` and `due`?

Reply:
good point, they can be dropped.